impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java

```diff
@@ -0,0 +1,241 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.dataservice.streaming.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import io.reactivex.Observable;
+import org.pentaho.di.core.RowMetaAndData;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleStepException;
+import org.pentaho.di.core.exception.KettleValueException;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.row.RowMetaInterface;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.dataservice.streaming.StreamList;
+import org.pentaho.di.trans.dataservice.utils.DataServiceConstants;
+import org.pentaho.di.trans.step.RowAdapter;
+import org.pentaho.di.trans.step.StepInterface;
+
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class represents a streaming execution for a service transformation.
+ * It spans a thread to run the transformation when data is requested and running thread exists, caching the requests.
+ * When all the cached requests are expired the transformation is stopped and it's associated thread terminated.
+ */
+public class StreamingServiceTransExecutor {
+  private final Trans serviceTrans;
+  private final String id;
+  private final String serviceStepName;
+  private final AtomicBoolean isRunning = new AtomicBoolean( false );
+
+  private StreamList<RowMetaAndData> stepStream;
+
+  /**
+   * Service listener cache.
+   */
+  private final Cache<String, StreamExecutionListener> serviceListeners = CacheBuilder.newBuilder()
+    .concurrencyLevel( 4 )
+    .expireAfterAccess( DataServiceConstants.STREAMING_CACHE_DURATION, DataServiceConstants.STREAMING_CACHE_TIME_UNIT )
+    .removalListener( new RemovalListener<String, StreamExecutionListener>() {
+      public void onRemoval( RemovalNotification<String, StreamExecutionListener> removal ) {
+        LogChannelInterface log = serviceTrans.getLogChannel();
+
+        removal.getValue().unSubscribe();
+
+        log.logDebug( DataServiceConstants.STREAMING_CACHE_REMOVED + removal.getKey() );
+      }
+    } )
+    .softValues()
+    .build();
+
+  private long lastCacheCleanupMillis = 0;
```

Comment:
is this lost in here?

Reply:
That's used to control the cleanup calls